src/main/scala/fi/aalto/cs/apluscourses/intellij/actions/ReplAction.scala

```diff
@@ -24,23 +26,62 @@ class ReplAction extends RunConsoleAction {
     customDoRunAction(e)
   }
 
-  def checkFileOrFolderIsNull(@Nullable fileOrFolder: VirtualFile): Boolean = fileOrFolder == null
+  def customDoRunAction(@NotNull e: AnActionEvent): Unit = {
+    val dataContext = e.getDataContext
+    val project = CommonDataKeys.PROJECT.getData(dataContext)
+    if (project == null) return // scalastyle:ignore
+
+    val runManagerEx = RunManagerEx.getInstanceEx(project)
+    val configurationType = getMyConfigurationType
+    val settings = runManagerEx.getConfigurationSettingsList(configurationType).asScala
+
+    /*
+     * The "priority order" is as follows:
+     *   1. If a file is open in the editor and it belongs to a module that has Scala SDK as a
+     *      library dependency, start the REPL for the module of the file.
+     *   2. Otherwise if a module (or file inside a module) is selected in the project menu on the
+     *      left and the module has Scala SDK as a library dependency, start the REPL for that
+     *      module.
+     *   3. Otherwise start a project level REPL
+     *
+     * Checking that the module has Scala SDK as a dependency is done to avoid the "no Scala facet
+     * configured for module" error.
+     */
+    val selectedModule = getScalaModuleOfEditorFile(project, dataContext)
+      .orElse(getScalaModuleOfSelectedFile(project, dataContext))
+
+    // Choose the configuration to run based on the condition if this a new configuration of not
+    val setting = settings.headOption.getOrElse {
+      val factory = configurationType.getConfigurationFactories.head
+      val configurationName = selectedModule
+        .map(module => s"REPL for ${module.getName}")
+        .getOrElse("Scala REPL")
+      runManagerEx.createConfiguration(configurationName, factory)
+    }
+
+    selectedModule match {
+      case Some(module) =>
+        val configuration = setting.getConfiguration.asInstanceOf[ScalaConsoleRunConfiguration]
+        if (setConfigurationConditionally(project, module, configuration)) {
+          RunConsoleAction.runExisting(setting, runManagerEx, project)
+        }
+      case None =>
+        val configuration = setting.getConfiguration.asInstanceOf[ScalaConsoleRunConfiguration]
+        configuration.setName("Scala REPL")
+        RunConsoleAction.runExisting(setting, runManagerEx, project)
+    }
+  }
```

Comment:
this looks.... _different_ :rofl: 
nice job, man, looks much more like the proper Scala code I've seen but unable to write myself :+1: